### DISCUSSION TITLE
Aidex CGMS support

### DISCUSSION THREAD

**Discussion by andyrozman** [NightscoutFoundation/xDrip#1977] [Q&A]
Are there any plans to support Aidex CGMS in near future? 

It seems it is the newcomer to UK and Ireland (Available in UK, not yet in IE). Prices also seem to be a nice competition to other system... Have no idea how accurate they are, but from what I have seen so far, they might be something for people who have to pay for CGMS themselves... and it seems that they will also be soon offered officially through Health insurance companies.

**Comment by ihewitt** [comment]
Good news! it should be fairly straightforward to add, I've managed to get data streaming direct from the sensor to a bangle.js watch (I find the javascript dev env makes BLE hacking childs play) [Picture](https://twitter.com/ivor_hewitt/status/1493335863211737091)
Still early days and obv needs to be converted to android, but I'll try and get some of the detail written up and sample code uploaded.

**Comment by andyrozman** [comment]
I started working on Android App a little, and there it also looks pretty straightforward, but creating pairing and stuff for primary connection seems little bit more involved than you described... But they have their own library and binary for communication, so I imagine that will handle most of the stuff. Maybe we could colaborate on this...
I should be getting my "hardware" by end of week I hope.

**Comment by ihewitt** [comment]
Yeah, I found the process for activating and pairing way too awkward... but I found it was possible to listen to the broadcast data and get the values out from that. So just activate and enable with the app and then listen for the data. e.g. some Espruino javascript code to do this:

[Gist link](https://gist.github.com/ihewitt/10a81c4bc98346b98cf4db2770b6aa15)

I expect it'll be a lot harder to connect and get historical data though.

**Comment by andyrozman** [comment]
So we managed to establish contact with Aidex company in China, and they are very eager to have support for their CGMS added to Xdrip. I hope we can get some support from their side, having permission to use their official CGMS library might get us there much faster, but we are still in beginning phases. Jon (jamorham) was also added to that communication thread, so that legal stuff could be discussed.

**Reply by ihewitt** [reply]
Excellent to hear (I had less luck when I tried!), I posted some more notes about the broadcast data I've been able to read out (and worked out a bit more of the format) [Link](https://blog.ivor.org/2022/03/sniffing-sugar.html)
Will be interesting to see what the library API is like and how it can be called, it did seem quite tightly coupled to their android app from looking at the code (i.e. dependant on registering callbacks into the native functions)

**Comment by andyrozman** [comment]
On Jon's suggestion I proposed to Aidex to just add broadcasting glucose data functionality to their official application. Since this would be very fast solution, they are quite eager to do this... Problem is they don't know how... I was looking through Xdrip code to see how Xdrip does broadcasting and I extracted that data into separate project. Would any of Xdrip developers be willing to take a look at this, just to see if my solution seems to be ok one. https://github.com/andyrozman/AidexSendXdrip

Thanks guys

**Comment by tzachi-dar** [comment]
I'll take a look on coming week.

**Comment by andyrozman** [comment]
I am currently talking to Aidex, I think they will extend their app to send broadcasts with data.

In the meantime I have testing app ready and working on reading data from Aidex CGMS. I am testing on my first sensor, with their library (not officially acquired) and their binary ble library, which sort of works... Its not perect but its working...

**Reply by keencave** [reply]
Great! Please keep us updated about the results!

**Reply by Bennani12** [reply]
Good work

**Comment by andyrozman** [comment]
Current status... So I am still talking with Aidex, they are keen to add broadcasting to their app but they are dragging their feet... I prepared test solutions for them how to send broadcast, and I added the same code to my small test application, and I added support for Aidex into AndroidAPS (just in my branch until we have final version of broadcasts when they add it into their App). I started working on adding support into Xdrip (my branch again: andyrozman:aidex_support_broadcast), but I have little bit problems with that... 

Would one of developers have little bit of time to give me some hints (I can already see values in graph, but I am missing following:
- value and slope is not displayed 
- I would like to display certain messages sometimes (like "Calibration requested", "Battery Low", etc)

**Reply by ihewitt** [reply]
nice, but how frustrating, I wonder if it's worth splitting the logic and using their "in the clear" broadcast data for the [live sensor feed](https://blog.ivor.org/2022/03/sniffing-sugar.html) and using the app integration for retrieving backlog/status notifications? 
Can't help with any xdrip knowledge/code questions I'm afraid. (and I'm waiting for another sensor)
I did spend some time to workout out the non-broadcast bluetooth protocol also and am quite close to having that decoded anyway. (just taking a little break from that for a while) I'll write up what I've got so far when I have any spare time.

**Comment by tzachi-dar** [comment]
As for displaying messages: There are two options: 1) on the Home class there are some version of the toast message. This displays things as a toast on the screen. This is useful when the user is looking at the screen. I don't think it will help with calibrations but might help with other things. For the battery there is already a place where it is displayed, you can use the same place (look for batter in the home class).

When you say value and slope not displayed:
1) Can you post a screen of what is displayed?
2) What hw data type did you choose?
3) What function are you calling to enter the BG into xDrip?

**Reply by Bennani12** [reply]
Any news about xdrip and aidex cgm

**Reply by andyrozman** [reply]
They (Aidex) implemented the first version, I tested it (with my xdrip branch), but there are few issues with it, so I sent information back to company, and I am waiting for next revision.

**Reply by qq254344379** [reply]
Very good news, can I download the test, I am wearing the sensor

**Reply by Bennani12** [reply]
I wear also now a sensor

**Reply by andyrozman** [reply]
The application I have is Chinese one, and I am not sure if they work with our transmitters.... or if you could upgrade your app to it (different id). You could contact me before you change sensor next and I can provide you with Aidex App, but like I said it has some errors in it... It reports raw value instead of app calculated one, sensorId is wrong, you can't configure where to send... (it sends to Xdrip and AAPS both).

**Reply by Bennani12** [reply]
I still have only 2 days in my sensor

**Reply by qq254344379** [reply]
I am Chinese, and I can understand Chinese. I've tested it to be compatible with different versions. I'm sure they're the same

**Reply by qq254344379** [reply]
I have tried different switching, but it is possible to switch between different software versions. Data can be synchronized. It can also be re-read

**Reply by qq254344379** [reply]
> 

My probe can be replaced, and I have multiple transmitters on my hands. Conveniently available under the software I downloaded to test

**Reply by qq254344379** [reply]
> The application I have is Chinese one, and I am not sure if they work with our transmitters.... or if you could upgrade your app to it (different id). You could contact me before you change sensor next and I can provide you with Aidex App, but like I said it has some errors in it... It reports raw value instead of app calculated one, sensorId is wrong, you can't configure where to send... (it sends to Xdrip and AAPS both).

Hello, I changed the sensor today, how do I get the broadcast app

**Comment by qq254344379** [comment]


**Reply by yoshimo** [reply]
That upload looks incomplete

**Comment by andyrozman** [comment]
I think that implementation for this is complete... What is process of getting it merged into main branch? Aidex App with support for Broadcasts should be out any day now (maybe it already is). I did last testing in begining of July... So any release after that might have it in.

**Reply by Bennani123** [reply]
Any update or news for Aidex and xdrip+

**Reply by Navid200** [reply]
> I think that implementation for this is complete... What is process of getting it merged into main branch? 

What changes to xDrip are required?
Have you made those changes in your fork?  

If yes, you can open a PR.
If no, how have you tested it?

**Comment by andyrozman** [comment]
I opened PR for this. As stated inside Aidex has now released app with broadcast worldwide... They have currently bug in it, which was reported to them and they are fixing it (they send differnt sensorId for each transmition, which breaks display of delta in Xdrip).

Branch was tested by me, and by at least one other user. 

Implementation is pretty straightforward... 

https://github.com/NightscoutFoundation/xDrip/pull/2312

**Reply by jiresongit** [reply]
Really looking forward to this, also happy to test if you need anyone more

**Comment by Bennani12** [comment]
Im a distributor of aidex in my country I still interested.  There is any update?

**Comment by jiresongit** [comment]
So I see this was merged last night and have downloaded. I've set the source, any other pertinent settings we need to apply? Not seeing anything coming thru either on the normal screen or readings yet? Am I asking in the right place

**Reply by Bennani12** [reply]
Have you  downloaded a new app ?

**Reply by jiresongit** [reply]
Yes and set sensor to glucorx aidex 
![Uploading Screenshot_2022-11-02-11-57-11-160-edit_com.eveningoutpost.dexdrip.jpg…]()

**Reply by Bennani12** [reply]
I can't see the screen shot.  Can you share me the new app please

**Reply by jiresongit** [reply]
It's in the nightly builds

**Reply by Bennani12** [reply]
Sorry have you a link . And thank you

**Reply by jiresongit** [reply]
https://github.com/NightscoutFoundation/xDrip/releases

**Reply by Bennani12** [reply]
Thank you . As you said no access to transmitter

**Reply by jiresongit** [reply]
@andyrozman are you able to give any.hints and tips please?

**Reply by andyrozman** [reply]
You need to configure Aidex as receiver and you need to have correct version of Aidex App installed... This is support for reading broadcasts received from Aidex App... Its not standalone solution. You still need to have their app installed.

**Reply by andyrozman** [reply]
You should be able to see something in the logs.. Filter by Aidex keyword and that should show you all the log messages...

**Reply by Bennani12** [reply]
Have you any information from technical service of Microtech about the adapted app ?

**Reply by andyrozman** [reply]
Don't understand the question? @Bennani12 
We (me) gave them instructions how to implement and they implemented it the way we wanted. Latest release was also tested with my branch... 
I have just asked the person that was testing it, to retest with released version...

**Reply by Bennani12** [reply]
Now I have Aidex  app working in my phone and connected with transmitter and sensor . I installed the new version of xdrip  and I change the source to Aidex app but no communication between the 2 app

**Reply by Bennani12** [reply]
Is this the right way?

**Reply by andyrozman** [reply]
In App you need set Location available always, this is important... Aidex App does BT scan on every reading and this requires Location to be set to ALWAYS_ON,,, Maybe that is the problem... (Yeah I know it is weird that it works that way)...

**Reply by Bennani12** [reply]
But still not working. Have you tested it ?

**Reply by andyrozman** [reply]
I didn't (I don't have Aidex anymore), but my tester did... And he said it works perfectly... 

Please join Discord channel, maybe people there can help... https://discord.gg/kRTQt8qNFx

**Comment by andyrozman** [comment]
So Aidex code was merged, but people are reporting problem, that they don't see the BG number (and arrows) visible... I am not sure if that is due to missconfiguration or is my implementation not correct. Could someone from Xdrip team check this? 

![image](https://user-images.githubusercontent.com/5500133/200794658-3f3212de-b540-4f16-96c2-101940ef2e7b.png)
